Fixes 210: added support for continous deployment from yaydoc

sch00lb0y opened the pull request:
<!--
Thanks for submitting a change to yaydoc!

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

-->

## Description
<!--- Describe your changes in detail -->
Integrated GitHub webhook to support continuous documentation deployment from yaydoc
## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
close #210 
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
right now we depending on third party to do continuous deployment. by having this feature which will enable to implement more feature as our wish in future and also no need for depending any third party ci.
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):
![ci](https://user-images.githubusercontent.com/10010900/27863315-39bb1ee6-61a7-11e7-9115-8a98e3352b01.png)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix 
- [X] New feature

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [Contribution & Best Practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [x] There is only one commit per issue.

mariobehling commented:
Please also resolve conflicts.

sch00lb0y commented:
@pri22296 @imujjwal96 please check this demo deployment http://52.221.209.153/ and I'll resolve conflicts once you reviewers okay with this PR 
Thanks.

pri22296 commented:
@sch00lb0y Please consider these points

- Remove permission to delete my repository
- Make the CI deploy button visible by default.

sch00lb0y commented:
@pri22296 now I added one more route for handling ci and I added `'write:repo_hook'` scope for permission. This is mandatory for to write the webhook. If it asks to delete repository permission we can't do anything for that. and CI deploy button is visible by default only. If this answer not solving your second query. please elaborate

pri22296 commented:
@sch00lb0y Sorry about my second query. Didn't realize that the button was there. My mistake.

sch00lb0y commented:
![github](https://user-images.githubusercontent.com/10010900/27839427-b76c194c-610e-11e7-85c8-3b9735cf4421.png)
@pri22296  It is not requesting any delete repo request. 
@imujjwal96 please review this PR.

pri22296 commented:
I tried it with my repo. Docs were not updated on push

sch00lb0y commented:
I have cleared the DB while checking that's why didn't work I guess. just now I tested with my test repo it is working fine.Please delete the webhook on your repository and register once again with yaydoc. it'll work for sure

pri22296 commented:
@sch00lb0y Checked it. worked this time

pri22296 commented on routes/ci.js:
```diff
@@ -0,0 +1,132 @@
+        var data = {
+          email: result.email,
+          gitUrl: req.body.repository.clone_url,
+          docTheme: "alabaster",
```
Don't hard code `alabaster`. Instead set it to be an empty string. `generate.sh` will handle the rest

sch00lb0y replied:
Made changes

pri22296 commented on routes/ci.js:
```diff
@@ -0,0 +1,132 @@
+    } else {
+      res.render("index", {
+        showMessage: true,
+        messages: `Yaydoc was already integrate with this repository`
```
Yaydoc is already integrated with this repository

sch00lb0y replied:
Made changes

pri22296 commented on routes/ci.js:
```diff
@@ -0,0 +1,132 @@
+          .then(function(result) {
+            res.render("index", {
+              showMessage: true,
+              messages: `Thanks for registering with Yaydoc.Here after Documentation will be pushed on each commit`
```
documentation will be pushed where? specify that.

sch00lb0y replied:
Made changes

pri22296 commented on routes/ci.js:
```diff
@@ -0,0 +1,132 @@
+          .catch(function(err) {
+            next({
+              status: 500,
+              message: 'something went wrong with db'
```
Something went wrong

sch00lb0y replied:
Made changes

imujjwal96 commented on backend/deploy.js:
```diff
@@ -1,6 +1,6 @@
 var crypter = require("../util/crypter.js")
 var spawn = require('child_process').spawn;
-
+var socketHandler = require('./socketHandler.js')
```
why is this required?

sch00lb0y replied:
now in the first parameter, you have to pass a socket object. I don't want to create as a separate function to handle generation. so I made a module named socket handler. it'll handle if the first parameter is socket otherwise it'll just ignore the object do nothing. we'll get a modular code

imujjwal96 replied:
When will it be undefined?

sch00lb0y replied:
check webhook handler. I'm just passing empty object
```js
            deploy.deployPages({}, {
              email: result.email,
              gitURL: req.body.repository.clone_url,
              username: result.username,
              uniqueId: generatedData.uniqueId,
              encryptedToken: result.accessToken
            })
```

imujjwal96 commented on model/repository.js:
```diff
@@ -0,0 +1,42 @@
+    new model({
+      name: name,
+      username: username,
+      githubId: githubId,
```
what's githubId?

sch00lb0y replied:
username can be changed so we can't use as a unique id. GitHub having uniqueId for each user. In order to sync with users, I'm using GitHub ID as GitHubId

imujjwal96 commented on routes/ci.js:
```diff
@@ -0,0 +1,132 @@
+
+router.post('/webhook', function(req, res, next) {
+  var event = req.get('X-GitHub-Event')
+  switch (event) {
```
why use switch case here?

sch00lb0y replied:
now, we use only push event. In future, we may use more webhook events so Instead of using if case I used switch

imujjwal96 commented:
instead of showing the message in a modal, show it in as a notification

imujjwal96 commented:
the flow is very vague. What i assumed (and certain that most users will assume) is that we have to fill the form before clicking on CI deploy. Think of a different flow

imujjwal96 commented:
What will happens if i have more than 100 repositories in my github profile

sch00lb0y commented:
>What will happens if i have more than 100 repositories in my github profile

since it is a single issue, making all feature in one PR is bit difficult after the PR merge I'll add search feature like Heroku.

> instead of showing the message in a modal, show it in as a notification

if any event triggers we can use notification, since we showing Thank you Message and Already registered Message I think Modal will be appropriate if we use notifiication it'll be unnoticed. 

> the flow is very vague. What i assumed (and certain that most users will assume) is that we have to fill the form before clicking on CI deploy. Think of a different flow

![ci](https://user-images.githubusercontent.com/10010900/27863168-7b7c0292-61a6-11e7-92b6-9d08b32d426f.png)
Now I have placed the button like this.is it okay ?

imujjwal96 commented:
The reason notifications exist is so that they are noticed.

imujjwal96 commented:
Do not headlining the form as WEBUI. The web UI consist of everything a user sees own our website, including the CI Deploy

sch00lb0y commented:
@imujjwal96 I removed modal and changed to the notification. I changed the headline to Web Generator. If you any other name in mind Please let me know

imujjwal96 commented:
Please resolve conflicts. I'm happy otherwise.

sch00lb0y commented:
@pri22296 @imujjwal96 I have resolved conflicts

pri22296 commented on backend/deploy.js:
```diff
@@ -19,15 +20,15 @@ exports.deployPages = function (socket, data) {
 
-  output.lineOutput(socket, process, 'deploy-logs', 18);
-  output.lineError(socket, process, 'err-logs');
+  socketHandler.handleLineOutput(socket, process, 'github-deploy-logs', 18);
```
Consider moving the logic of `socketHandler` to util/output.js

sch00lb0y replied:
I moved `socketHandler.js` to util folder. but moving logic to `output.js` won't make much sense.

pri22296 replied:
They seem to be doing similar things that's why I asked to move it If it was possible. Anyways seems fine.

pri22296 commented on backend/generator.js:
```diff
@@ -4,11 +4,11 @@ var mailer = require('./mailer');
+var socketHandler = require('./socketHandler.js');
 
-exports.executeScript = function (socket, formData) {
+exports.executeScript = function (socket, formData, cb) {
```
I suppose `cb` stands for `callback`? If that is the case, consider naming it `callback`. shouldn't be a lot of extra keystrokes.

sch00lb0y replied:
just a js standard where people widely use. https://stackoverflow.com/questions/31780872/what-is-cb-in-node Anyways I made changes

pri22296 commented:
@sch00lb0y Please merge your changes from upstream

pri22296 commented:
@sch00lb0y Just a tip. In future, use `git pull --rebase upstream master` to avoid creating unnecessary merge commits.

imujjwal96 commented:
@sch00lb0y Can you show us a Heroku deployment with this change?

sch00lb0y commented:
@imujjwal96 please check this http://demo08.herokuapp.com/

imujjwal96 commented:
![image](https://user-images.githubusercontent.com/8245662/27949423-48a3df04-631b-11e7-9b61-f35c88063193.png)
**S**elect the repository
**R**egister

imujjwal96 commented:
why's the route /ci/register
![image](https://user-images.githubusercontent.com/8245662/27949640-0632d642-631c-11e7-9879-9cd1e870b65e.png)

sch00lb0y commented:
`why's the route /ci/register` POST request for form submission. but then I can use '/'  but it won't look good and also it'll become difficult to understand the code

imujjwal96 commented:
Why do I perform a basic documentation generation at the /ci/register route? This doesn't look good and frankly doesn't make sense.

sch00lb0y commented:
@imujjwal96 I'll change the route to POST `/`

imujjwal96 commented:
Oh please don't do this. What I'm saying is that there is a particular design pattern that web application follows. 
We first send a **P**OST request, perform an action, **R**edirect and then render a template in a **G**ET request.
https://en.wikipedia.org/wiki/Post/Redirect/Get

Want you've done is that you're rendering the index page in the POST request itself. What this will lead to is that every time the user refreshes that page, the POST request is sent again. This is something that is never desired.


What I want you to do is that instead of rendering the index page in you POST request route, redirect to the `/` route with a registration successful notification

sch00lb0y commented:
@imujjwal96 Yeah I already made changes according to that I'll test and let you know soon

imujjwal96 commented on routes/ci.js:
```diff
@@ -0,0 +1,126 @@
+  },
+  function (error, response, body) {
+    var repos = JSON.parse(body);
+    res.render("index", {
```
I think we should keep the index template specific for the `/` route. What is happening here is that since the modal is collapsible, i can perform the actions which are aesthetically supposed to be performed in the `/` route, at the `/ci/register` route.

sch00lb0y replied:
Can I do the same `res.redirect("/?message=Thanks for registering with Yaydoc.Hereafter Documentation will be pushed to the GitHub pages on each commit.");` but the logic for index route will become big. if you have any suggestion please do give me

imujjwal96 replied:
do res.redirect("/?success=true") and do something similar to what i did at ownCloud
https://github.com/owncloud/core/pull/27277/files#diff-5af29c9f0fdb4338457f3bb7024dc991R152

imujjwal96 replied:
That is, remove the query params as soon as the server reads it and shows notification

sch00lb0y replied:
@imujjwal96 I passing the repo array in query string and changing url state . please check https://demo08.herokuapp.com

imujjwal96 commented on views/index.jade:
```diff
@@ -86,5 +96,33 @@ block content
+    $("#ciModal").modal('show');
+  if showMessage === true
+   script.
+     $('.notification').append($('<li>')).text("#{message}");
```
use `styles.showNotification()` function defined in `public/scripts/styles.js`

sch00lb0y replied:
made changes

imujjwal96 commented on routes/ci.js:
```diff
@@ -0,0 +1,125 @@
+          crypter.encrypt(req.session.token),
+          req.session.email)
+          .then(function(result) {
+            res.redirect("/?message=Thanks for registering with Yaydoc.Hereafter Documentation will be pushed to the GitHub pages on each commit.");
```
instead of spitting out the message as a get request, use `/?status=true` or something like '/?status=200' 
and in `index.jade` line 119 check for this status and display the message

sch00lb0y replied:
made changes

sch00lb0y commented:
@imujjwal96 I have updated the branch.Please check

mariobehling commented:
http://52.221.209.153/ links to localhost when I click on CI.

sch00lb0y commented:
@mariobehling I have updated the README.md . Please check the readme [here](https://github.com/sch00lb0y/yaydoc/tree/continuous). I was testing in localhost that's why it is redirected to localhost. I have changed that also. check the deployment here https://demo08.herokuapp.com/ . 

Since it it is a big PR not able to add this feature 
>  Add info how user can choose theme and change location/branch of documentation site

I'll do this feature after this pr get merged

mariobehling commented:
Ok, thanks. I guess theme changes etc. can be taken care of in following issues. Cheers.
